### PR TITLE
TDigest

### PR DESCRIPTION
**Pull request by AlexanderSaydakov**

### CODE REVIEW

File: src/test/resources/tdigest_ref_k100_n10000_double.sk
[the remark this answers is not in the document]
Comment (AlexanderSaydakov) [reply src/test/resources/tdigest_ref_k100_n10000_double.sk:1]: that is generated by the reference implementation

File: src/main/java/org/apache/datasketches/tdigest/TDigestDouble.java [hunk 1]
```diff
@@ -0,0 +1,636 @@
...
+   * @return normalized rank (from 0 to 1 inclusive)
+   */
+  public double getRank(final double value) {
+    if (isEmpty()) throw new SketchesStateException(QuantilesAPI.EMPTY_MSG);
+    if (Double.isNaN(value)) throw new SketchesArgumentException("Operation is undefined for Nan");
+    if (value < minValue_) return 0;
+    if (value > maxValue_) return 1;
+    if (numCentroids_ + numBuffered_ == 1) return 0.5;
+
+    mergeBuffered(); // side effect
```
[the remark this answers is not in the document]
Comment (AlexanderSaydakov) [reply src/main/java/org/apache/datasketches/tdigest/TDigestDouble.java:193]: we never noted such things before. I think it is an implementation detail

File: src/main/java/org/apache/datasketches/tdigest/TDigestDouble.java [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+    int sizeBytes = preambleLongs * Long.BYTES +
+        (isEmpty() ? 0 : (isSingleValue() ? Double.BYTES : 2 * Double.BYTES + (Double.BYTES + Long.BYTES) * numCentroids_));
+    final byte[] bytes = new byte[sizeBytes];
+    final WritableBuffer wbuf = WritableMemory.writableWrap(bytes).asWritableBuffer();
+    wbuf.putByte(preambleLongs);
+    wbuf.putByte(SERIAL_VERSION);
+    wbuf.putByte((byte) Family.TDIGEST.getID());
+    wbuf.putShort(k_);
+    wbuf.putByte((byte) (
+      (isEmpty() ? 1 << flags.IS_EMPTY.ordinal() : 0) |
```
[the remark this answers is not in the document]
Comment (AlexanderSaydakov) [reply src/main/java/org/apache/datasketches/tdigest/TDigestDouble.java:307]: Not sure. We rely on implicit values all over the place.

File: src/main/java/org/apache/datasketches/tdigest/TDigestDouble.java [hunk 3]
```diff
@@ -0,0 +1,636 @@
...
+    maxValue_ = max;
+    if (k < 10) throw new SketchesArgumentException("k must be at least 10");
+    int fudge = 0;
+    if (USE_WEIGHT_LIMIT) {
+      fudge = 10;
+      if (k < 30) fudge += 20;
+    }
+    centroidsCapacity_ = k_ * 2 + fudge;
+    bufferCapacity_ = centroidsCapacity_ * 5;
+    double scale = Math.max(1.0, (double) bufferCapacity_ / centroidsCapacity_ - 1.0);
```
[the remark this answers is not in the document]
Comment (AlexanderSaydakov) [reply src/main/java/org/apache/datasketches/tdigest/TDigestDouble.java:503]: This is from the reference implementation. Perhaps we could take some liberties to simplify.

File: src/main/java/org/apache/datasketches/tdigest/TDigestDouble.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmalkin) [reply src/main/java/org/apache/datasketches/tdigest/TDigestDouble.java:195]: will calling this repeatedly change the result? if yes, that seems important to no.  if no, then i agree it's an implementation detail.

**Review by jmalkin — APPROVED** [review]
I know checkstyle would complain about the javadocs.  Pretty much anything else is an optimizaiton.

**Pull request merged**